impress-org/givewp
# Donors can now update their credit card using the Donor Dashboard

Resolves #5901

## Description

This PR resolves the issue where donors were not able to update their subscriptions and credit card details using the Donor Dashboard. 

**How update credit card details work for Stripe subscriptions?** 

In order to update the card details for a subscription using Stripe, the following order of events has to happen
1. User enters valid credit card details
2. An request to stripe servers is executed to get a new payment id based on new credit card details
3. Another request is executed where we pass the new payment id and subscription id to update the subscription details

The problem was that step 3. was executed before step 2. and we couldn't pass the new payment id because we didn't have it yet. 

The issue is resolved by blocking the **update request** until we get the new payment id. After the new payment id is obtained, the update request is executed.


## Visuals
![Screen Capture_select-area_20210730142401](https://user-images.githubusercontent.com/4222590/127652550-97ddd275-80ac-422f-bf5c-1361d873527d.gif)


## Testing Instructions


1. Install GiveWP, Recurring Donations and Stripe premium
2. Should have the donor dashboard on the website
3. Create a form with recurring donations enabled
4. Make a donation using stripe credit card
5. Access the donor dashboard and change your credit card
6. Check on your Stripe account and see if the card is updated


## Pre-review Checklist
